assets/training/finetune_acft_hf_nlp/environments/acpt/context/finetune_run.py

```diff
@@ -367,32 +368,109 @@ def add_task_specific_params(cmd: List[str], task_name: str, component_name: str
                     add_optional_input(cmd, param)
 
 
-def _run_subprocess_cmd(cmd: List[str], component_name: str, completion_files_folder: str):
+def is_main_process():
+    """
+    To check if current process running this is master or rank_0 process.
+
+    :return: Boolean for whether this process is master.
+    """
+    return os.environ.get('AZUREML_PROCESS_NAME', 'main') in {'main', 'rank_0'}
+
+
+def wait_at_barrier(barrier_file, num_processes):
+    """
+    Control will halt till number to process reaching the execution point is less than a given number.
+
+    barrier_file: File used to create execution barrier.
+    num_processes: Number of process which need to reach barrier point.
+    """
+    with open(barrier_file, 'a+') as f:
+        f.seek(0)
+        lines = f.readlines()
+        process_count = len(lines)
+        process_name = os.environ.get('AZUREML_PROCESS_NAME', 'main')
+        logger.info(f'Process {process_name} at barrier, count is {process_count}, in barrier file {barrier_file}')
+        if process_count < num_processes:
+            f.write(f"{os.getpid()} reached the barrier\n")

```

Comment:
There seems to be a potential race condition here. Think about it as discussed:  if process 1 opens the file as of lines 388/389, proceeds to finish executing line 393 then gets context switched. At this point lets say num_processes is 3 and the file desc points to the EOF which is line 4.
Now process 2 starts executing the function, runs through to completion, exhausting its time slice of execution. It also gets num_processes as 3 and goes on to write its pid at line 4.
If process 1 gets scheduled back again, now it starts executing at line 394 - so will write its own pid and overwrite the previous write.
Isn't that a classic case of race condition at the MP level?
This can be solved by a guard bit (use another guard file) to access this file with the pid.

Reply:
Thinking to use https://pypi.org/project/portalocker/ for file based lock. It gives portability across windows and Linux with MP support

Reply:
I am not familiar with portalocker. But if it seems reasonable, please try it. Raise a PR with the new code or usage then. Will take a look at the new implementation then.